### ISSUE TITLE
Shrink RuleNode by two words

### ISSUE THREAD

**Issue by bholley** [servo/servo#17280]
RuleNode is a frequently-allocated data structure. It lives in components/style/rule_tree/mod.rs

Right now it's 96 bytes on 64-bit platforms. We can shrink this to 80 quite easily.

We should do the following:
(1) Add a size_of test for RuleNode in tests/unit/stylo/size_of.rs.
(2) Change StyleSource to have a "None" variant, and eliminate the Option<> around StyleSource.
(3) Reuse a sibling pointer for free_count, as described in the comment.

CC @nox @jdm @emilio @SimonSapin

**Comment by emilio** [comment]
>  (2) Change StyleSource to have a "None" variant, and eliminate the Option<> around StyleSource.

This wouldn't shrink anything, AFAICT, would it? StyleSource contains an `Arc`, which contains `NonZero`.

**Comment by bholley** [comment]
> This wouldn't shrink anything, AFAICT, would it? StyleSource contains an Arc,
> which contains NonZero.

It would, because the NonZero optimization doesn't extend into nested enums: https://is.gd/lviB8t

I did indeed verify that Option<StyleSource> is 24 bytes.

**Comment by zuwow** [comment]
I'm interested in working on this.
@highfive: assign me

**Comment by bholley** [comment]
Great! Let me know if you have any questions.

**Comment by zuwow** [comment]
I'm not sure what to do about the read function on the StyleSource struct.
https://github.com/servo/servo/blob/master/components/style/rule_tree/mod.rs#L96
With the addition of a None variant, what should block be? Or should that be handled at the caller like it is now with the option?

**Comment by bholley** [comment]
Right - it should just panic, and the callers should check is_none().

**Comment by zuwow** [comment]
Makes sense. To be clear then, you're thinking that StyleSource would look like
```rust
enum StyleSource  {
    Style(...),
    Declarations(...),
    None
}
```
and then I'd add an `is_none()` function to StyleSource . Any trait I should use for that?

Sorry, my rust is a bit uh... rusty.

**Comment by jdm** [comment]
No need for a trait. You can use a bare `impl StyleSource` to add the method.

**Comment by zuwow** [comment]
I was able to remove the option and refactor everything such that RuleNode is now 88 bytes and passing the unit tests. Though I don't know what to do for ptr_equals in the case where they're both None since there's no reference to check. Have it return true, leave it as is and have it be false, or have it panic?

For free_count though I'm a little confused. First, I'm not sure what your proposed fix from the comment would entail but I probably don't have enough context on this whole module. Second, I don't think free_count is doing anything. None of the usages I could find in the code involved modifying its value.

```
❯ git grep -n free_count
components/style/rule_tree/mod.rs:546:    free_count: AtomicUsize,
components/style/rule_tree/mod.rs:568:            free_count: AtomicUsize::new(0),
components/style/rule_tree/mod.rs:583:            free_count: AtomicUsize::new(0),
components/style/rule_tree/mod.rs:913:        me.free_count.store(0, Ordering::Relaxed);
components/style/rule_tree/mod.rs:920:        if self.get().free_count.load(Ordering::Relaxed) > RULE_TREE_GC_INTERVAL {
```

When it was first added ([commit](https://github.com/servo/servo/commit/4e52bb49b9fc7850cf75c5af8950538b14b856bc#diff-149c665e272df5ba8e0ce4ec55e53e42R567)), it was being modified but that line was removed [here](https://github.com/servo/servo/commit/e67ea42c3fa03264b8c897430f7c17a7c153e5bd#diff-149c665e272df5ba8e0ce4ec55e53e42L574). It looks like it might be possible to simply remove it.

**Comment by bholley** [comment]
> Have it return true, leave it as is and have it be false, or have it panic?

It should panic, and the callers should check for None.

> Second, I don't think free_count is doing anything.

Wow, nice find! That basically means we'll never GC the rule tree until the document is torn down, which is really bad. I'll fix that, and handle the sibling-pointer-reuse thing while I'm at it. So just focus on the StyleSource thing.

**Comment by bholley** [comment]
Filed [1] on the free count thing.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1373725

**Comment by bholley** [comment]
Landed the fix for the other stuff in #17368.

**Comment by zuwow** [comment]
Created pull request #17404 for this.

I ended up creating `is_some()` instead of `is_none()` since that made the code that calls it more clear (no need to do `! is_none()`).
